src/utils/leak.js

```diff
@@ -586,7 +586,337 @@ export class LeakImplementation {
+    ): void {
+      let environmentBindings = record.bindings;
+      for (let bindingName of Object.keys(environmentBindings)) {
+        let binding = environmentBindings[bindingName];
+        invariant(binding !== undefined);
+        let found = nonLocalReadBindings.delete(bindingName);
+        // Check what undefined could mean here, besides absent binding
+        // #2446
+        if (found && binding.value !== undefined) {
+          if (binding.hasLeaked) {
+            notSupportedForTransitiveMaterialization();
+          }
+          computeFromValue(binding.value);
+        }
+      }
+    }
+    function computeFromAbstractValue(value: AbstractValue): void {
+      if (value.values.isTop()) {
+        if (value.kind === "conditional") {
+          computeFromValue(value.args[1]);
+          computeFromValue(value.args[2]);
+        } else {
+          let whitelistedKind =
+            value.kind &&
+            (value.kind === "widened numeric property" || // TODO: Widened properties needs to be havocable.
+              value.kind.startsWith("abstractCounted"));
+          invariant(
+            whitelistedKind || value.intrinsicName || value.args.length > 0,
+            "Havoced unknown object requires havocable arguments"
+          );
+
+          // TODO: This is overly conservative. We recursively leak all the inputs
+          // to this operation whether or not they can possible be part of the
+          // result value or not.
+          for (let arg of value.args) {
+            computeFromValue(arg);
+          }
+        }
+      } else {
+        // If we know which object this might be, then leak each of them.
+        for (let element of value.values.getElements()) {
+          computeFromValue(element);
+        }
+      }
+    }
+    function computeFromProxyValue(value: ProxyValue): void {
+      computeFromValue(value.$ProxyTarget);
+      computeFromValue(value.$ProxyHandler);
+    }
+    function computeFromValue(value: Value): void {
+      if (value.isIntrinsic() || value instanceof EmptyValue || value instanceof PrimitiveValue) {
+        visit(value);
+      } else if (value instanceof AbstractValue) {
+        ifNotVisited(value, computeFromAbstractValue);
+      } else if (
+        value.isIntrinsic() &&
+        value instanceof ArrayValue &&
+        ArrayValue.isIntrinsicAndHasWidenedNumericProperty(value)
+      ) {
+        ifNotVisited(value, computeFromObjectValue);
+      } else if (value instanceof FunctionValue) {
+        ifNotVisited(value, computeFromFunctionValue);
+      } else if (value instanceof ObjectValue) {
+        ifNotVisited(value, computeFromObjectValue);
+      } else if (value instanceof ProxyValue) {
+        ifNotVisited(value, computeFromProxyValue);
+      }
+    }
+    function computeFromObjectValue(value: Value): void {
+      invariant(value instanceof ObjectValue);
+      if (!objectsToMaterialize.has(value)) objectsToMaterialize.add(value);
```

Comment:
What about arrays and functions which are objects too?

Reply:
We shouldn't have to add values that have already leaked (and have thus been materialized and all environment interactions go through generator entries).

Reply:
Lumped into the larger #2485

Reply:
But how does that function-object case work when we don't add function values to the objects to materialize?

Reply:
Could you elaborate what you mean by "don't add function values to the objects to materialize?" An example would help. The main outer function (i.e. the array op) is passed to `computeFromFunctionValue` which in turn explores its properties.